File: app/infrastructure/models/listing_model.py
import datetime
import enum
import uuid
from typing import Optional

from sqlalchemy import DateTime, Enum, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import Mapped, mapped_column

from app.infrastructure.models.base_model import Base


class TransactionType(enum.Enum):
    SELL = "Sell"
    RENT = "Rent"


class PropertyType(enum.Enum):
    HOUSE = "House"
    APARTMENT = "Apartment"


class Status(enum.Enum):
    AVAILABLE = "Available"
    PENDING = "Pending"
    CLOSED = "Closed"


class Listing(Base):
    __tablename__ = "listings"
    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, index=True, default=uuid.uuid4
    )
    client_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("clients.id"),
        nullable=True,
    )
    user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id"), nullable=True
    )
    title: Mapped[str] = mapped_column(unique=True)
    location: Mapped[str] = mapped_column()
    street: Mapped[str] = mapped_column()
    price: Mapped[int] = mapped_column()
    area: Mapped[float] = mapped_column()
    property_type: Mapped[PropertyType] = mapped_column(Enum(PropertyType))
    description: Mapped[str] = mapped_column()
    transaction_type: Mapped[TransactionType] = mapped_column(Enum(TransactionType))
    floor: Mapped[str] = mapped_column()
    num_of_floors: Mapped[str] = mapped_column()
    build_year: Mapped[str] = mapped_column()
    status: Mapped[Status] = mapped_column(Enum(Status), nullable=False)
    created_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(), server_default=func.current_timestamp(), nullable=False
    )

    @hybrid_property
    def price_per_area(self) -> float:
        if self.area and self.area != 0:
            return round(self.price / self.area)
        return 0.0

File: frontend/src/app/core/components/listings/listing/listing.model.ts
export interface Listing extends ListingInput {
  price_per_area: number;
  id: string;
  created_at: Date;
}

export interface ListingInput {
  client_id: string;
  title: string;
  location: string;
  street: string;
  price: number;
  area: number;
  property_type: propertyType;
  description: string;
  transaction_type: transactionType;
  floor: string;
  num_of_floors: string;
  build_year: string;
  status: status;
  user_id?: number;
}

export enum transactionType {
  sell = 'Sell',
  rent = 'Rent',
}

export enum propertyType {
  house = 'House',
  apartment = 'Apartment',
}

export enum status {
  available = 'Available',
  pending = 'Pending',
  closed = 'Closed',
}
